### PR TITLE
Try new Rust grammar

### PR DESCRIPTION
**Pull request by alexr00**
Related to https://github.com/microsoft/vscode/issues/64488

### CODE REVIEW

File: extensions/rust/test/colorize-results/test_rs.json [hunk 1]
```diff
@@ -154,19 +242,30 @@
 		}
 	},
 	{
-		"c": "println!",
-		"t": "source.rust support.function.std.rust",
+		"c": "println",
```
Comment (alexr00) [inline extensions/rust/test/colorize-results/test_rs.json:245]: In other languages (and in the old grammar for Rust), `print` statements are `support.function` should that be the case for Rust too?

File: extensions/rust/test/colorize-results/test_rs.json [hunk 2]
```diff
@@ -210,7 +342,29 @@
...
+		"r": {
+			"dark_plus": "default: #D4D4D4",
+			"light_plus": "default: #000000",
+			"dark_vs": "default: #D4D4D4",
+			"light_vs": "default: #000000",
+			"hc_black": "default: #FFFFFF"
+		}
+	},
+	{
+		"c": "mut",
```
Comment (alexr00) [inline extensions/rust/test/colorize-results/test_rs.json:366]: In the old grammar `mut` was `storage.modifier`, is `keyword` more correct?

File: extensions/rust/test/colorize-results/test-6611_rs.json [hunk 3]
```diff
@@ -781,7 +847,678 @@
...
+			"dark_plus": "default: #D4D4D4",
+			"light_plus": "default: #000000",
+			"dark_vs": "default: #D4D4D4",
+			"light_vs": "default: #000000",
+			"hc_black": "default: #FFFFFF"
+		}
+	},
+	{
+		"c": "struct",
+		"t": "source.rust keyword.control.rust",
```
Comment (alexr00) [inline extensions/rust/test/colorize-results/test-6611_rs.json:1104]: In other languages, `struct` is scoped as `storage.type`, and in the old grammar it was also `storate.type`. Should it change?

File: extensions/rust/test/colorize-results/test-6611_rs.json [hunk 4]
```diff
@@ -781,7 +847,678 @@
...
+		"r": {
+			"dark_plus": "default: #D4D4D4",
+			"light_plus": "default: #000000",
+			"dark_vs": "default: #D4D4D4",
+			"light_vs": "default: #000000",
+			"hc_black": "default: #FFFFFF"
+		}
+	},
+	{
+		"c": "trait",
```
Comment (alexr00) [inline extensions/rust/test/colorize-results/test-6611_rs.json:1301]: Should this stay `storage.type`?

**Comment by alexr00** [comment]
@dustypomerleau I have tried adopting your rust grammar in this PR. We have colorization tests that compare the color and scope of tokens in sample language files. The language files for rust are here: 
https://github.com/Microsoft/vscode/blob/5bb9d0b0987c9db4335535e7b339a26d20dda473/extensions/rust/test/colorize-fixtures/test-6611.rs#L1-L1
https://github.com/Microsoft/vscode/blob/8509d4e47a8e85355353500d0b0636d3fcbd0462/extensions/rust/test/colorize-fixtures/test.rs#L1-L1

Your grammar definitely colors more tokens and colors some tokens more specifically, which is great. However, there are some tokens that I'm not confident have the correct scope. I left some comments in the output of the colorization tests where the scope doesn't match what I'd expect.

As you probably know already, the reason we have some standard scopes that we expect various tokens to be is so that one theme can look good on any language. I am not a rust user, so I'm not 100% certain that the comments I have are relevant, but please take a look and convince me that the scopes are correct (or update your grammar and I'll pull the change into this PR).

Once we've addressed the scope differences, I'll bring this grammar into insiders for a month or two. Given the low user count for this grammar currently, I want to make sure that it has plenty of time in insiders so users and find any issues.

File: extensions/rust/test/colorize-results/test_rs.json (hunk 1 above)
Comment (bjorn3) [reply extensions/rust/test/colorize-results/test_rs.json:245]: In Rust `println!` is a macro and not a function. I also think that `println!` maybe should be considered a single token and not two tokens (`println` and `!`)

File: extensions/rust/test/colorize-results/test-6611_rs.json (hunk 3 above)
Comment (bjorn3) [reply extensions/rust/test/colorize-results/test-6611_rs.json:1104]: https://macromates.com/manual/en/language_grammars

> storage — things relating to “storage”.
>     type — the type of something, class, function, int, var, etc.

Yeah, it should be `storage.type` I think.

File: extensions/rust/test/colorize-results/test_rs.json [hunk 5]
```diff
@@ -210,7 +342,29 @@
 	},
 	{
 		"c": "let",
```
[the remark this answers is not in the document]
Comment (bjorn3) [reply extensions/rust/test/colorize-results/test_rs.json:344]: I think so. Even if not, `keyword.control` seems wrong as it doesn't affect control flow.

File: extensions/rust/test/colorize-results/test_rs.json (hunk 2 above)
Comment (bjorn3) [reply extensions/rust/test/colorize-results/test_rs.json:366]: > * storage — things relating to “storage”.
>     * modifier — a storage modifier like static, final, abstract, etc.

Should be `storage.modifier` I think.

**Comment by dustypomerleau** [comment]
@alexr00 The reason I chose `keyword` for many of these scopes is because that's how they are classified by the Rust lexer. Strict keywords listed in [The Rust Reference](https://doc.rust-lang.org/reference/keywords.html#strict-keywords) include most of the tokens caught by your test:

- `impl`
- `let`
- `mut`
- `struct`
- `trait`

I'm happy to move some of the keywords into `storage`, but which ones? Anything that allocates on the heap?

As for `println!`, @bjorn3 has already mentioned that this is a macro in rust. It is one of the [macros provided by the standard library](https://github.com/rust-lang/rust/blob/cc662cd4540c6d3b8cb14cf535fa023699503dff/library/std/src/macros.rs#L93), so I can see how it might be viewed as `support`. My only opinion here is that it should be possible to theme macros that use the `macro_bang!` syntax differently from plain functions. I personally don't feel a need to have different scopes for standard library macros versus user-defined macros, but if others felt strongly I could separate them. As for the `!` itself, I can provide a separate scope for that (distinct from it's use as NOT), and let theme designers decide whether to match it to the rest of the macro or let it stand out.

**Comment by alexr00** [comment]
@dustypomerleau since many of the tokens that get scoped as `storage` in other languages are also keywords in their respective languages, I think something like the guideline that @bjorn3 quoted is a good rule of thumb for when to use `storage`: https://github.com/microsoft/vscode/pull/108254#discussion_r501675868

What you said about macros makes sense to me. 

Please ping here when you've made the changes you want to make and I'll update again!

**Comment by dustypomerleau** [comment]
@alexr00 Changes have been pushed. Appreciate your help.

**Comment by dustypomerleau** [comment]
@alexr00 Pushed an additonal commit to fix a highlighting error for raw strings https://github.com/rust-analyzer/rust-analyzer/issues/6042.

**Comment by alexr00** [comment]
@dustypomerleau I'll try to get to this this week. Thanks for the updates!

**Comment by alexr00** [comment]
I was not able to review this again for October. Will address in November.

**Comment by alexr00** [comment]
@dustypomerleau I am happy with the changes, thanks! I will merge the grammar from https://github.com/dustypomerleau/rust-syntax/ into VS Code and we will see what users think for a month.

**Comment by dustypomerleau** [comment]
Thanks for your help, @alexr00! For patch updates, should I ping someone or submit my own PR with updated commit hash? We are still finding occasional edge cases since I merged into Rust Analyzer.

**Comment by alexr00** [comment]
@dustypomerleau if there's an urgent fix in the grammar, please just open an issue and ping me on it. Otherwise, I pick up grammar changes with a script about once per month at the beginning of our milestones.

**Comment by alexr00** [comment]
and @bjorn3 thanks for the additional feedback!

File: extensions/rust/test/colorize-results/test_rs.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (cynecx) [reply extensions/rust/test/colorize-results/test_rs.json:322]: The new grammer has an issue I am seeing right now:

`fn` and `use` as should not be highlighted as `keyword.control`. In my opinion `fn` does not not necessarily signify control-flow. The same applies to `use`, there is no control-flow.

**Comment by cynecx** [comment]
The grammer in it's current state has some issues, I'd like to get fixed:

> The new grammer has an issue I am seeing right now:
> 
> `fn` and `use` as should not be highlighted as `keyword.control`. In my opinion `fn` does not not necessarily signify control-flow. The same applies to `use`, there is no control-flow.

Is it preferred to open a PR here or upstream (https://github.com/dustypomerleau/rust-syntax/). Also the current situation is quite weird, because we've got this same grammar in multiple places vscode, upstream extension and in rust-analyzer.

**Comment by bjorn3** [comment]
I think vscode doesn't accept any changes to grammars. I think they only periodically pull in updates from the upstream version, which is https://github.com/dustypomerleau/rust-syntax/ in this case.

**Comment by alexr00** [comment]
@bjorn3 is correct.

**Comment by alexr00** [comment]
There has been no feedback about this grammar having problems. I will leave it in for the release of 1.52.

**Pull request merged**